### PR TITLE
feat: Generalize the UDAFs collect_list and collect_set

### PR DESCRIPTION
**Pull request by jnh5y**
Addresses https://github.com/confluentinc/ksql/issues/7697

### Description 
The goal is to allow the UDAFs collect_list and collect_set to operate on arbitrary input types.  

### Testing done 
QTTs have been added to show the functions working generally.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/array/CollectListUdaf.java [hunk 1]
```diff
@@ -47,54 +51,51 @@ private CollectListUdaf() {
...
+  public static <T> TableUdaf<T, List<T>, List<T>> createCollectListT() {
+    return new Collect<>();
+  }
+
   public static TableUdaf<Long, List<Long>, List<Long>> createCollectListLong() {
     return new Collect<>();
   }
 
-  @UdafFactory(description = "collect values of an Integer field into a single Array")
   public static TableUdaf<Integer, List<Integer>, List<Integer>> createCollectListInt() {
```
Comment (jnh5y) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/array/CollectListUdaf.java:62]: These functions are unneeded now.  That said, they are public.  What do folks think about removing them?

Comment (agavra) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/array/CollectListUdaf.java:62]: yeah we can remove them, nobody should be calling these directly

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/collect-set.json [hunk 2]
```diff
@@ -8,6 +8,86 @@
     "a join or another aggregate."
   ],
   "tests": [
+    {
+      "name": "collect_set with arrays, structs, and maps",
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/collect-set.json:12]: can we add tests for both of these for decimal type as well (and nested decimal). We should make sure that the set membership check works. 

Also what if we have a struct with a `null` value in one of the fields (we should make sure the set membership check works there as well)? What about a `null` at the top level?

**Review by agavra — APPROVED** [review]
Nice! You can definitely remove all the `@UdafFactory` methods. Nobody should be calling them publicly and it'll clean up a lot of code. Love to see it.

**Pull request merged**